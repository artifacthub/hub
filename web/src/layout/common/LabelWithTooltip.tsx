import isNull from 'lodash/isNull';
import isUndefined from 'lodash/isUndefined';
import React, { useEffect, useState } from 'react';

import styles from './LabelWithTooltip.module.css';

interface Props {
  active?: boolean | null;
  label: JSX.Element;
  visibleTooltip: boolean;
  tooltipMessage: string;
  className?: string;
}

const LabelWithTooltip = (props: Props) => {
  const [visibleTooltipStatus, setVisibleTooltipStatus] = useState<boolean>(false);
  const [onLabelHover, setOnLabelHover] = useState<boolean>(false);

  useEffect(() => {
    let timeout: NodeJS.Timeout;
    if (!visibleTooltipStatus && onLabelHover) {
      timeout = setTimeout(() => {
        setVisibleTooltipStatus(true);
      }, 200);
    }
    if (visibleTooltipStatus && !onLabelHover) {
      timeout = setTimeout(() => {
        setVisibleTooltipStatus(false);
      }, 50);
    }
    return () => {
      if (!isUndefined(timeout)) {
        clearTimeout(timeout);
      }
    };
  }, [onLabelHover, visibleTooltipStatus]);

  if (isUndefined(props.active) || isNull(props.active) || !props.active) return null;

  return (
    <div className={props.className}>
      <div
        data-testid="labelWithTooltip"
        onMouseEnter={(e) => {
          e.preventDefault();
          setOnLabelHover(true);
        }}
        onMouseLeave={() => {
          setOnLabelHover(false);
        }}
      >
        {props.label}
      </div>

      {props.visibleTooltip && visibleTooltipStatus && (
        <div className="position-absolute">
          <div className={`tooltip bs-tooltip-bottom ${styles.tooltip}`} role="tooltip">
            <div className={`arrow ${styles.tooltipArrow}`} />
            <div className={`tooltip-inner ${styles.tooltipContent}`}>{props.tooltipMessage}</div>
          </div>
        </div>
      )}
    </div>
  );
};

export default LabelWithTooltip;
