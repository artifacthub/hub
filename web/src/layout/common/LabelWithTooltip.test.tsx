import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';

import LabelWithTooltip from './LabelWithTooltip';

const defaultProps = {
  tooltipMessage: 'Tooltip message',
  label: <div>label</div>,
  active: true,
  visibleTooltip: true,
};

describe('LabelWithTooltip', () => {
  it('creates snapshot', () => {
    const { asFragment } = render(<LabelWithTooltip {...defaultProps} />);
    expect(asFragment).toMatchSnapshot();
  });

  it('renders properly', () => {
    const { getByTestId, getByText } = render(<LabelWithTooltip {...defaultProps} />);
    expect(getByTestId('labelWithTooltip')).toBeInTheDocument();
    expect(getByText('label')).toBeInTheDocument();
  });

  it('does not render label', () => {
    const props = {
      ...defaultProps,
      active: false,
    };
    const { container } = render(<LabelWithTooltip {...props} />);
    expect(container).toBeEmpty();
  });

  it('displays tootltip', () => {
    const { getByTestId, getByText, getByRole } = render(<LabelWithTooltip {...defaultProps} />);

    const badge = getByTestId('labelWithTooltip');
    fireEvent.mouseEnter(badge);

    waitFor(() => {
      expect(getByText(defaultProps.tooltipMessage)).toBeInTheDocument();
      expect(getByRole('tooltip')).toBeInTheDocument();
    });
  });

  it('hides tootltip on mouse leave', () => {
    const { getByTestId, getByText, getByRole, queryByRole } = render(<LabelWithTooltip {...defaultProps} />);

    const badge = getByTestId('labelWithTooltip');
    fireEvent.mouseEnter(badge);

    waitFor(() => {
      expect(getByText(defaultProps.tooltipMessage)).toBeInTheDocument();
      expect(getByRole('tooltip')).toBeInTheDocument();
    });

    fireEvent.mouseLeave(badge);

    waitFor(() => {
      expect(queryByRole('tooltip')).toBeNull();
    });
  });

  it('does not display tootltip when visibleTooltip is false', () => {
    const props = {
      ...defaultProps,
      visibleTooltip: false,
    };
    const { getByTestId, queryByRole } = render(<LabelWithTooltip {...props} />);

    const badge = getByTestId('labelWithTooltip');
    fireEvent.mouseEnter(badge);

    waitFor(() => {
      expect(queryByRole('tooltip')).toBeNull();
    });
  });
});
