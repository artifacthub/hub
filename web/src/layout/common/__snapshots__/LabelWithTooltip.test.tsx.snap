// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelWithTooltip creates snapshot 1`] = `[Function]`;
